File: src/levels/level_0.zig
const std = @import("std");
const log = std.log;
const Allocator = std.mem.Allocator;
const assert = std.debug.assert;

const ECS = @import("ecs");

const graphics = @import("../graphics.zig");
const mod = @import("../mod.zig");
const tile = @import("../tile.zig");
const Object = @import("../object.zig");
const Player = @import("../player.zig");
const Sight = @import("../sight.zig");
const TileMemory = @import("../tile_memory.zig");
const Scene = @import("../scene.zig");
const Level = @import("../scenes/level.zig");

const mainspace = @import("../main.zig");
const nc = mainspace.nc;
const sdl = mainspace.sdl;

// Remove extra tiles after this if possible
const maxTiles = 400;

pub var level: Level = .{
  .id = .Level0,
  .allocator = undefined,
  .scene = .{.id = .Level, .vtable = sceneVTable},
  .vtable = vtable,
};

const sceneVTable = Scene.VTable{
  .init = struct {fn init(allocator: Allocator) !*const Scene
  {
    level.allocator = allocator;

    return &level.scene;
  }}.init,

  .enter = struct {fn enter(self: *const Scene) !*const Scene
  {
    _ = self;
    return &level.scene;
  }}.enter,

  .getInput = struct {fn getInput(
    self: *const Scene,
    inputEvent: sdl.SDL_Event) !void
  {
    _ = self;
    _ = inputEvent;
  }}.getInput,

  .update = struct {fn update(self: *const Scene) !void {
    _ = self;

    const playerSight =
      mainspace.ecs.getPtr(Level.objects.items[0].id, "sight", Sight).?;
    var keys = level.tiles.keys();
    for (keys) |key|
    {
      if (level.tiles.count() <= maxTiles)
      {
        break;
      }

      if (!playerSight.inView(key))
      {
        log.debug("Remove tile {}\n", .{
          key - mainspace.ecs.getComponent(
            Level.objects.items[0].id, "pos", Level.Coord
          ).?
        });
        if (!level.tiles.orderedRemove(key)) unreachable;
        keys = level.tiles.keys();
      }
    }
  }}.update,

  .draw = struct {fn draw(self: *const Scene)
    (error{TileNotFound} || graphics.Error)!void
  {
    _ = self;
  }}.draw,

  .exit = struct {fn exit(self: *const Scene) !void {_ = self;}}.exit,

  .deinit = struct {fn deinit(self: *const Scene) !void
  {
    _ = self;
    level.deinit();
  }}.deinit,
};

const vtable = Level.VTable{
  .generateTile = struct {fn generateTile(self: *Level, pos: Level.Coord)
    Allocator.Error!ECS.Entity.Unmanaged
  {
    _ = self;
    _ = pos;

    //if (result == 0)
    //{
    //  result = mainspace.ecs.addEntity(.{
    //    .tileType = tile.nameTypes.get(
    //      .{.mod = "base", .name = "cyanideCarpet"}
    //    ).?,
    //  }).id;
    //}

    //if (@mod(pos[0], 2) == 1 and @mod(pos[1], 2) == 1)
    //{
    //  result = mainspace.ecs.addEntity(.{
    //    .tileType = tile.nameTypes.get(
    //      .{.mod = "base", .name = "cyanideCarpet"}
    //    ).?,
    //  }).id;
    //} else if (@mod(pos[0], 2) == 0 and @mod(pos[1], 2) == 0)
    //{
    //  result = mainspace.ecs.addEntity(.{
    //    .tileType = tile.nameTypes.get(
    //      .{.mod = "base", .name = "yellowWallpaper"}
    //    ).?,
    //  }).id;
    //} else
    //{
    //  if (mainspace.rand.uintLessThan(u2, 3) == 0)
    //  {
    //    result = mainspace.ecs.addEntity(.{
    //      .tileType = tile.nameTypes.get(
    //        .{.mod = "base", .name = "yellowWallpaper"}
    //      ).?,
    //    }).id;
    //  } else
    //  {
    //    result = mainspace.ecs.addEntity(.{
    //      .tileType = tile.nameTypes.get(
    //        .{.mod = "base", .name = "cyanideCarpet"}
    //      ).?,
    //    }).id;
    //  }
    //}

    //if (@rem(pos[0], 2) == 0 and @rem(pos[1], 2) == 0 and
    //  tile.getStaticData(
    //    try self.getTile(Level.Coord{pos[0]-1, pos[1]})
    //  ).?.walkable and
    //  tile.getStaticData(
    //    try self.getTile(Level.Coord{pos[0]+1, pos[1]})
    //  ).?.walkable and
    //  tile.getStaticData(
    //    try self.getTile(Level.Coord{pos[0], pos[1]-1})
    //  ).?.walkable and
    //  tile.getStaticData(
    //    try self.getTile(Level.Coord{pos[0], pos[1]+1})
    //  ).?.walkable)
    //{
    //  mainspace.ecs.getComponentPtr(result, "tileType", tile.Type).?.* =
    //    tile.nameTypes.get(
    //      .{.mod = "base", .name = "cyanideCarpet"}
    //    ).?;
    //}

    //return result;
    return 0;
  }}.generateTile,

  .getCamPos = struct {fn getCamPos(self: Level) Level.Coord
  {_ = self;
    return Level.getCenterCameraOn(Level.objects.items[0].id) catch unreachable;
  }}.getCamPos,
};

File: mods/base/levels/level_0.lua
-- The functions and tables defined in this file are automatically namespaced, and as such are an exception to the no globals rule
-- self is an alias for fractal.mods.modName.levels.levelName
init = {
  name = "level0"
}

function deinit(self)

end

function enter(self)
  -- Remove extra tiles after this if possible
  self.tiles.max = 400
  
  print("Hello, Level 0!");
  --try Level.objects.append(init.gpa, .init(0, .{0, 0}, .{
  --  .sight = Sight{.radius = 15, .view = .empty},
  --  .tileMemory = TileMemory{.tiles = .empty},
  --}));
  --try Turn.push(init.gpa, &ecs, Level.objects.getLast());
  --defer Turn.queue.deinit(init.gpa);

  --defer ecs.getPtr(
  --  Level.objects.items[0].id, "tileMemory", TileMemory
  --).?.tiles.deinit(init.gpa);
  --defer ecs.getPtr(
  --  Level.objects.items[0].id, "sight", Sight
  --).?.view.deinit(init.gpa);

  -- Oh, no! It's not lore accurate!
  self.objects:add(
    {"base", "smiler"},
    {pos = {30, 30}}
  )

end

function exit(self)

end

local function removeExtraTiles(self)
  self.tiles.max = 400

  local player = fractal.mods.base.player
  local playerPos = player.pos:get()

  for k, _ in self.tiles:iterate() do
    if self.tiles:count() <= self.tiles.max then
      break
    end

    -- Use approximation for distance since this could be done several times per turn
    local dis = math.abs(k[1]-playerPos[1]) + math.abs(k[2]-playerPos[2])
    if dis > 10 and not player.sight:inView(k) then
      print("Remove tile { ", k[1], ", ", k[2], " }")
      self.tiles:remove(k)
      -- May need to resync iterator at this point
    end
  end
end

function update(self)
  self.camera:centerOn(fractal.mods.base.player)

  --local array = {1, 1, 2, 3, 5, 8, 13, 21}
  --local function iterator(array)
  --  local index = 0
  --  return 
  --   function()
  --     index = index + 1
  --     if index > #array then
  --       return nil
  --     else
  --       return array[index]
  --     end
  --   end
  --end

  --for element in iterator(array) do
  --  print(element)
  --end

  removeExtraTiles(self)
end

function generateTile(self, pos)
  --return fractal.mods.base.levels.level1:generateTile(pos)
  local originDis = math.abs(pos[1]) + math.abs(pos[2])

  local result = nil
  if pos[1]%2 == 1 and pos[2]%2 == 1 then
    result = {"base", "cyanideCarpet"}
  elseif pos[1]%2 == 0 and pos[2]%2 == 0 then
    result = {"base", "yellowWallpaper"}
  else
    --if originDis > 4 and math.random(0, 3) == 0 then
    --if math.random(0, math.max(3, 32-math.floor(originDis))) == 0 then
    if math.random(0, 3) == 0 then
      result = {"base", "yellowWallpaper"}
    else
      result = {"base", "cyanideCarpet"}
    end
  end

  if pos[1]%2 == 0 and pos[2]%2 == 0 and
    self.tiles:get({pos[1]-1, pos[2]}):typeData().walkable and
    self.tiles:get({pos[1]+1, pos[2]}):typeData().walkable and
    self.tiles:get({pos[1], pos[2]-1}):typeData().walkable and
    self.tiles:get({pos[1], pos[2]+1}):typeData().walkable
  then
    result = {"base", "cyanideCarpet"}
  end

  --print("Generate "..result[1].." at {"..tostring(pos).."}")
  return result
end

